examples/enc-search-sample.rs

```diff
@@ -0,0 +1,421 @@
+        .await?;
+    let dec_name = std::str::from_utf8(&dec_result.decrypted_data()).unwrap();
+    let dec_name_trans = ironoxide::search::transliterate_string(&dec_name);
+    if name_parts
+        .iter()
+        .all(|name_part| dec_name_trans.contains(name_part))
+    {
+        Ok(Some(dec_name.to_string()))
+    } else {
+        Ok(None)
+    }
+}
+// end-snippet{filterCust}
+
+// Given a query string, generate the set of index tokens and use this to
+// retrieve possible matches from the server. Transliterate the string,
+// break it into pieces on white space, and for each returned customer,
+// check to see if the customer name actually contains the words from the
+// query. If so, output the customer ID and name.
+// start-snippet{displayCust}
+async fn display_matching_customers(
+    sdk: &IronOxide,
+    name_index: &BlindIndexSearch,
+    query_str: &str,
+) -> Result<()> {
+    let query_tokens = name_index
+        .tokenize_query(query_str, None)?
+        .into_iter()
+        .collect();
+    let customer_recs = search_customers(&query_tokens);
+    let trans_query = ironoxide::search::transliterate_string(&query_str);
+    let name_parts: Vec<&str> = trans_query.split_whitespace().collect();
+    for cust in customer_recs.iter() {
+        let result = filter_customer(&sdk, &cust, &name_parts).await?;
+        match result {
+            Some(decrypted_name) => println!("{} {} matched query", cust.id, decrypted_name),
+            None => println!("{} did not match query", cust.id),
+        }
+    }
+    Ok(())
+}
+// end-snippet{displayCust}
+
+// Mock out some functions that would call the back-end service and return data in the real app
+// We use some mutable global state to persist things that the service usually would.
+struct GlobalState {
+    pub encrypted_salt: String,
+    pub customers: Vec<Customer>,
+    pub start_time: String,
+}
+
+impl GlobalState {
+    pub fn new() -> Self {
+        GlobalState {
+            encrypted_salt: "".to_string(),
+            customers: vec![],
+            start_time: SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap()
+                .as_secs()
+                .to_string(),
+        }
+    }
+}
+
+lazy_static! {
+    static ref GLOBAL_STATE: MutStatic<GlobalState> = MutStatic::new();
+}
+
+fn save_encrypted_salt_to_app_server(encrypted_salt: String) {
+    GLOBAL_STATE.write().unwrap().encrypted_salt = encrypted_salt;
+}
+
+fn get_encrypted_salt_from_app_server() -> String {
+    GLOBAL_STATE.read().unwrap().encrypted_salt.clone()
+}
+
+// For now, we ignore the tokens associated with each customer rec
+fn save_customer(customer: Customer, _name_tokens: &[u32], _email_tokens: &[u32]) {
+    GLOBAL_STATE.write().unwrap().customers.push(customer);
+}
+
+// Just return all the records that we have in the store for now
+fn search_customers(_tokens: &Vec<u32>) -> Vec<Customer> {
```

Comment:
If we're trying to make this look similar to a real application it seems like we shouldn't just return all the values. This causes all the records they've added to always be returned giving the impression that there's no way to filter before you decrypt. 

To fix this all we have to do is change the state to be a `Map<Vec<u32>, Customer>>` and filter based on all the tokens matching.

Reply:
Used HashSets for the tokens.